core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -1122,8 +1141,45 @@ wasmtime_ssp_fd_renumber(wasm_exec_env_t exec_env, struct fd_table *curfds,
     fd_object_release(exec_env, fo);
     --ft->used;
 
+    // Handle renumbering of any preopened resources
+    struct fd_prestat *prestat_from;
+    __wasi_errno_t prestat_from_error =
+        fd_prestats_get_entry(prestats, from, &prestat_from);
+
+    struct fd_prestat *prestat_to;
+    __wasi_errno_t prestat_to_error =
+        fd_prestats_get_entry(prestats, to, &prestat_to);
+
+    // Renumbering over two preopened resources.
+    if (prestat_from_error == 0 && prestat_to_error == 0) {
+        error = fd_prestats_remove_entry(prestats, to);
+
+        if (error == 0) {
+            error = fd_prestats_insert_locked(prestats, prestat_from->dir, to);
+
+            if (error == 0) {
+                error = fd_prestats_remove_entry(prestats, from);
+            }
```

Comment:
Same as above, should we restore data changed by `fd_prestats_remove_entry` when `error != 0`?